clang/tools/checked-c-convert/ArrayBoundsInferenceConsumer.cpp

```diff
@@ -1,127 +1,573 @@
+  if (commonSubSeqLen > 0) {
+    // Check if we get 80% match on the common subsequence matching on the
+    // variable name of length and the name of array.
+    return ((commonSubSeqLen * 100.0) / (ptrName.length() * 1.0)) >=
+           COMMONSUBSEQUENCEPERCMATCH;
+  }
+  return false;
+
+}
+
+static bool fieldNameMatch(std::string lenFieldName) {
+  // Convert the field name to lower case.
+  std::transform(lenFieldName.begin(), lenFieldName.end(), lenFieldName.begin(),
+                 [](unsigned char c){ return std::tolower(c); });
+  for (auto &potentialName : possibleLengthVarNamesPrefixes) {
+    if (lenFieldName.rfind(potentialName, 0) == 0)
+      return true;
+  }
+
+  for (auto &potentialName : possibleLengthVarNamesSubstring) {
+    if (lenFieldName.find(potentialName) != std::string::npos)
+      return true;
+  }
+  return false;
+}
+
+static bool hasLengthKeyword(std::string varName) {
+  // Convert the field name to lower case.
+  std::transform(varName.begin(), varName.end(), varName.begin(),
+                 [](unsigned char c){ return std::tolower(c); });
+
+  std::set<std::string> allLengthKeywords(possibleLengthVarNamesPrefixes);
+  allLengthKeywords.insert(possibleLengthVarNamesSubstring.begin(),
+                           possibleLengthVarNamesSubstring.end());
+  for (auto &potentialName : allLengthKeywords) {
+    if (varName.find(potentialName) != std::string::npos)
+      return true;
+  }
+  return false;
+}
+
+// Check if the provided constraint variable is an array and it needs bounds.
+static bool needArrayBounds(ConstraintVariable *CV,
+                            Constraints::EnvironmentMap &envMap) {
+  if (CV->hasArr(envMap)) {
+    PVConstraint *PV = dyn_cast<PVConstraint>(CV);
+    if (PV && PV->getArrPresent())
+      return false;
+    return true;
+  }
+  return false;
+}
+
+static bool needNTArrayBounds(ConstraintVariable *CV,
+                              Constraints::EnvironmentMap &envMap) {
+  if (CV->hasNtArr(envMap)) {
+    PVConstraint *PV = dyn_cast<PVConstraint>(CV);
+    if (PV && PV->getArrPresent())
+      return false;
+    return true;
+  }
+  return false;
+}
+
+static bool needArrayBounds(Decl *decl, ProgramInfo &Info, ASTContext *Context) {
+  std::set<ConstraintVariable*> consVar = Info.getVariable(decl, Context);
+  for (auto currCVar: consVar) {
+    if (needArrayBounds(currCVar, Info.getConstraints().getVariables()))
+      return true;
+    return false;
+  }
+  return false;
+}
+
+// Map that contains association of allocator functions and indexes of
+// parameters that correspond to the size of the object being assigned.
+static std::map<std::string, std::set<int>> AllocatorSizeAssoc = {
+                                            {"malloc", {0}},
+                                            {"calloc", {0, 1}}};
+
+
+// Get the name of the function called by this call expression.
+std::string getCalledFunctionName(Expr *E) {
```

Comment:
To prevent accidental changes to function arguments consider making them const. And if a function does not change any of its arguments consider making the function const. This can be done throughout your code.